### PR TITLE
Add support for generic EFI booting

### PR DESCRIPTION
**Pull request by jakogut**
Depends-on: https://github.com/balena-os/balena-generic/pull/14
https://github.com/balena-os/meta-balena/pull/2085

### CODE REVIEW

**Comment by alexgg** [comment]
@resin-jenkins add to whitelist

**Comment by jakogut** [comment]
@alexgg, What does "add to whitelist" do? Is this documented somewhere?

**Comment by alexgg** [comment]
> @alexgg, What does "add to whitelist" do? Is this documented somewhere?

@jakogut it's the Jenkins github PR builder whitelisting, see https://plugins.jenkins.io/ghprb/

**Comment by alexgg** [comment]
@jakogut I have sorted out the jenkins jobs, please rebase over master to resolve the conflict and make sure the PR builds for the existing device types `generic--aarch64` and `generic-armv7ahf`.

**Comment by alexgg** [comment]
@balena-ci rebase

**Comment by jakogut** [comment]
@alexgg I waited to rebase this one until the dunfell branch it depends on was merged. I just pushed it rebased onto origin/master.

File: layers/meta-balena-generic/recipes-kernel/linux/linux-yocto_%.bbappend [hunk 1]
```diff
@@ -0,0 +1,6 @@
+inherit kernel-resin
+
+KCONFIG_MODE="--alldefconfig"
+COMPATIBLE_MACHINE_generic-aarch64 = "generic-aarch64"
+KBUILD_DEFCONFIG_generic-aarch64 ?= "defconfig"
+
```
Comment (alexgg) [inline layers/meta-balena-generic/recipes-kernel/linux/linux-yocto_%.bbappend:8]: Would be interesting to see the difference between the .config used here for amd64 and the one from the `genericx86-64-ext` DT in balena-intel that we want to replace with this.

**Review by alexgg — COMMENTED** [review]
On device repositories we need one or more "Changelog-entry" and "Change-type" is not needed.

File: generic-amd64.coffee [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	gettingStartedLink:
+		windows: 'https://docs.resin.io/learn/getting-started/generic-amd64/nodejs/#add-your-first-device'
+		osx: 'https://docs.resin.io/learn/getting-started/generic-amd64/nodejs/#add-your-first-device'
+		linux: 'https://docs.resin.io/learn/getting-started/generic-amd64/nodejs/#add-your-first-device'
+
+	supportsBlink: false
+
+	yocto:
+		machine: 'generic-amd64'
+		image: 'resin-image'
```
[the remark this answers is not in the document]
Comment (klutchell) [reply generic-amd64.coffee:30]: use this branch of meta-balena:
https://github.com/balena-os/meta-balena/tree/klutchell/resin-img-rename

and balena-yocto-scripts `v1.11.0` or newer to test against this renamed stuff

File: generic-aarch64.coffee [hunk 3]
```diff
@@ -28,5 +28,14 @@ module.exports =
 	yocto:
 		machine: 'generic-aarch64'
 		image: 'resin-image'
+		fstype: 'resinos-img'
```
[the remark this answers is not in the document]
Comment (klutchell) [reply generic-aarch64.coffee:31]: See above for testing resin/balena renames. Though there are more device repo changes required for compatibility.

File: generic-amd64.coffee [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+DOCKER_IMAGE = 'Download/pull the following docker image on the target machine: <code>{{ dockerImage }}</code> .'
+DOWNLOAD_CONFIG = 'Download the configuration file and deploy it on target.'
+RUN_CONTAINER = 'Run the resinOS container using <a href="https://github.com/resin-os/resinos-in-container">resinos-in-container</a>. This tool/script will reference the image and the configuration file (downloaded above) as command line arguments. Example: <code>./resinos-in-container.sh --config ./config.json --image {{ dockerImage }}</code>. Tweak the <code>--config</code> argument based on the filename set for the downloaded configuration file.'
+
+module.exports =
+	version: 1
+	slug: 'generic-amd64'
+	name: 'Generic AMD64 (x86-64)'
+	arch: 'amd64'
+	state: 'experimental'
```
[the remark this answers is not in the document]
Comment (jakogut) [reply generic-amd64.coffee:13]: @alexgg Should this be changed to `state: 'new'`?

File: generic-amd64.coffee [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+	instructions: [
+		DOCKER_IMAGE
+		DOWNLOAD_CONFIG
+		RUN_CONTAINER
```
[the remark this answers is not in the document]
Comment (jakogut) [reply generic-amd64.coffee]: @alexgg Would it make sense to show how to boot these images using QEMU? The process of importing or writing the image to a server is going to be specific to the platform or cloud provider.

[the remark this answers is not in the document]
Comment (alexgg) [reply generic-amd64.coffee]: qemu is not a real device. I think generic instructions to run these on an x86_64 workstation/server are better. Not something like EC2, but just a standard x86_64 machine.

File: generic-amd64.coffee (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexgg) [reply generic-amd64.coffee:13]: Probably. We will need to discuss in an arch/product call how the UEFI device type will replace the generic-x86_64-ext.

**Comment by mtoman** [comment]
@resin-jenkins test this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
@resin-jenkins add to whitelist

**Comment by alexgg** [comment]
@balena-ci retest this please

File: layers/meta-balena-generic/recipes-kernel/linux/linux-yocto_%.bbappend (hunk 1 above)
Comment (mtoman) [reply layers/meta-balena-generic/recipes-kernel/linux/linux-yocto_%.bbappend:8]: This does not block the PR, issue created https://github.com/balena-os/balena-generic/issues/20

File: generic-amd64.coffee (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtoman) [reply generic-amd64.coffee]: Changed to `new`, created a ticket to track the future/replacement: https://github.com/balena-os/balena-generic/issues/21

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by jakogut** [comment]
@balena-ci test

**Comment by jakogut** [comment]
@balena-ci retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

**Comment by mtoman** [comment]
@resin-jenkins retest this please

File: generic-amd64.coffee [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	]
+
+	gettingStartedLink:
+		windows: 'https://docs.resin.io/learn/getting-started/generic-amd64/nodejs/#add-your-first-device'
+		osx: 'https://docs.resin.io/learn/getting-started/generic-amd64/nodejs/#add-your-first-device'
+		linux: 'https://docs.resin.io/learn/getting-started/generic-amd64/nodejs/#add-your-first-device'
```
[the remark this answers is not in the document]
Comment (mtoman) [reply generic-amd64.coffee]: Docs issue created - https://github.com/balena-io/docs/issues/1903
I have mentioned that the docs can probably wait until we figure out https://github.com/balena-os/balena-generic/issues/21

**Review by mtoman — APPROVED** [review]

**Pull request merged**